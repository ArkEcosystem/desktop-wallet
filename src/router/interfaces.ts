import { Environment } from "@arkecosystem/platform-sdk-profiles";
import { Location } from "history";

export interface MiddlewareParams {
	location: Location;
	env: Environment;
	redirect: (url: string) => void;
}

export interface Middleware {
	handler(params: MiddlewareParams): boolean;
}
