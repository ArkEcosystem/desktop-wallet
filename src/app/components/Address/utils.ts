/**
 * Truncates a given string adding
 * dots in the middle (e.g AUdQs...SfdQs).
 */
export const truncateStringMiddle = (input: string, maxChars = 20) => {
	if (input.length <= maxChars) return input;

	const midPos = Math.floor(maxChars / 2) - 2;
	const start = input.substr(0, midPos);
	const end = input.substr(input.length - midPos, input.length);
	return `${start}...${end}`;
};
