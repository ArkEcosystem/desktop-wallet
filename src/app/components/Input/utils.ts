type RangeInput = {
	min: number;
	max: number;
	step: number;
};

const isStepDivisible = ({ min, max, step }: RangeInput) => {
	const res = (max - min) / step;
	return parseInt(res.toString(), 10) === res;
};

export const sanitizeStep = ({ min, max, step }: RangeInput): number =>
	isStepDivisible({ min, max, step }) ? step : (max - min) / 100;
