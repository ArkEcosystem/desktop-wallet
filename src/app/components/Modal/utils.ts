export const modalOffsetClass = (itemHeight: number, windowHeight: number) => {
	const padding = 80;

	return itemHeight + 2 * padding > windowHeight ? "top-0 my-20" : "";
};
