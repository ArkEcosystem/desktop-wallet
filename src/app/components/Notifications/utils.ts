import { Profile } from "@arkecosystem/platform-sdk-profiles";

export const markAsRead = async (isVisible: boolean, id: string, profile: Profile, env: any) => {
	if (!isVisible) {
		return;
	}

	const notification = profile.notifications().get(id);
	if (!notification || notification?.read_at) {
		return;
	}

	profile.notifications().markAsRead(id);
	await env.persist();
};
