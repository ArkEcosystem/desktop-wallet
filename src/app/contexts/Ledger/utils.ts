import { BigNumber } from "@arkecosystem/platform-sdk-support";

export type LedgerData = {
	index: number;
	address: string;
	path: string;
	balance?: BigNumber;
	isNew?: boolean;
	timestamp?: number;
};

export type LedgerDerivationScheme = {
	coinType: number;
	purpose?: number;
	account?: number;
	change?: number;
	address?: number;
};

export const customDerivationModes = {
	ARK: {
		legacy: (slip44: number, cursor: number) => formatLedgerDerivationPath({ coinType: slip44, account: cursor }),
		bip44: (slip44: number, cursor: number) => {
			/** The legacy mode also starts with this path, so it doesn't need to be scanned again */
			if (cursor === 0) {
				return;
			}

			return formatLedgerDerivationPath({ coinType: slip44, address: cursor });
		},
	},
};

export const formatLedgerDerivationPath = (scheme: LedgerDerivationScheme) =>
	`${scheme.purpose || 44}'/${scheme.coinType}'/${scheme.account || 0}'/${scheme.change || 0}/${scheme.address || 0}`;
