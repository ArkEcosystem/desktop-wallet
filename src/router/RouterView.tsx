import React from "react";
import { RouteConfig } from "react-router-config";
import { Route, Switch } from "react-router-dom";
import { styled } from "twin.macro";

const Wrapper = styled.div``;

type Props = {
	routes: RouteConfig[];
	wrapper?: React.ElementType;
};

export const RouterView = ({ routes, wrapper }: Props) => {
	return (
		<Switch>
			{routes.map((route, i) => (
				<Route
					key={i}
					path={route.path}
					render={(props) => (
						<Wrapper data-testid="RouterView__wrapper" as={wrapper}>
							{/* @ts-ignore */}
							<route.component {...props} routes={route.routes} />
						</Wrapper>
					)}
				/>
			))}
		</Switch>
	);
};
