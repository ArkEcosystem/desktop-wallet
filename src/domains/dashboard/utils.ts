import { Profile, ProfileSetting } from "@arkecosystem/platform-sdk-profiles";
import { uniq } from "@arkecosystem/utils";

export const enableNetworkInDashboardFilters = (profile: Profile, networkId: string) => {
	const config: any = profile.settings().get(ProfileSetting.DashboardConfiguration) || { selectedNetworkIds: [] };
	config.selectedNetworkIds.push(networkId);
	profile.settings().set(ProfileSetting.DashboardConfiguration, {
		...config,
		selectedNetworkIds: uniq(config.selectedNetworkIds),
	});
};
