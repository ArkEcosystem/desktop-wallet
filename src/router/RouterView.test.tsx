/* eslint-disable react/display-name */
import { createMemoryHistory } from "history";
import React from "react";
import { MemoryRouter, Router, withRouter } from "react-router-dom";
import { render } from "utils/testing-library";

import { Middleware, MiddlewareParams } from "./interfaces";
import { RouterView } from "./RouterView";

describe("RouterView", () => {
	const LocationDisplay = withRouter(({ location }) => (
		// @ts-ignore
		<div data-testid="location-display">{location.location?.pathname || location.pathname}</div>
	));

	it("should render", () => {
		const { getByTestId, asFragment } = render(
			<MemoryRouter>
				<RouterView routes={[{ path: "/", component: () => <h1>Test</h1> }]} />
			</MemoryRouter>,
		);
		expect(getByTestId("RouterView__wrapper")).toHaveTextContent("Test");
		expect(asFragment()).toMatchSnapshot();
	});

	it("should render with custom wrapper", () => {
		const { getByTestId, asFragment } = render(
			<MemoryRouter>
				<RouterView wrapper="section" routes={[{ path: "/", component: () => <h1>Test</h1> }]} />
			</MemoryRouter>,
		);
		expect(getByTestId("RouterView__wrapper").tagName).toEqual("SECTION");
		expect(asFragment()).toMatchSnapshot();
	});

	it("should block /test router", () => {
		const handler = jest.fn(({ location }: MiddlewareParams) => location.pathname !== "/test");
		const testMiddleware: Middleware = {
			handler,
		};

		const history = createMemoryHistory();
		history.push("/test");

		const { getByTestId } = render(
			<Router history={history}>
				<RouterView
					routes={[{ path: "/test", component: () => <h1>Test</h1> }, { component: () => <h1>Home</h1> }]}
					middlewares={[testMiddleware]}
				/>
			</Router>,
		);

		expect(handler).toHaveBeenCalledTimes(2);
		expect(getByTestId("RouterView__wrapper")).toHaveTextContent("Home");
	});

	it("should block /test router and redirect to a custom url", () => {
		const handler = jest.fn(({ location, redirect }: MiddlewareParams) => {
			if (location.pathname === "/test") {
				redirect("/custom");
				return false;
			}
			return true;
		});

		const testMiddleware: Middleware = {
			handler,
		};

		const history = createMemoryHistory();
		history.push("/test");

		const { getByTestId } = render(
			<Router history={history}>
				<>
					<RouterView
						routes={[
							{ path: "/test", component: () => <h1>Test</h1> },
							{ path: "/custom", component: () => <h1>Custom</h1> },
						]}
						middlewares={[testMiddleware]}
					/>
					<LocationDisplay />
				</>
			</Router>,
		);

		expect(handler).toHaveBeenCalledTimes(2);
		expect(getByTestId("location-display")).toHaveTextContent("/custom");
	});
});
