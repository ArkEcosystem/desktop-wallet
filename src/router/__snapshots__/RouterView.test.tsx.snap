// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouterView should render 1`] = `
<DocumentFragment>
  <div
    class="RouterView__Wrapper-sef01e-0"
    data-testid="RouterView__wrapper"
  >
    <h1>
      Test
    </h1>
  </div>
</DocumentFragment>
`;

exports[`RouterView should render with custom wrapper 1`] = `
<DocumentFragment>
  <section
    class="RouterView__Wrapper-sef01e-0"
    data-testid="RouterView__wrapper"
  >
    <h1>
      Test
    </h1>
  </section>
</DocumentFragment>
`;
